More encodings

kahaaga opened the pull request:
## Summary

After this PR, we have the following `Encoding`s:
- `OrdinalPatternEncoding`
- `GaussianCDFEncoding` (extended to state vectors in this PR, was just for integers before)
- `RectangularBinEncoding`
- `AmplitudeEncoding` (new)
- `FirstDifferenceEncoding` (new)

Any combination (of two or more) of these five encodings is also a valid encoding (by using the new `CombinationEncoding`). 
Thus, we will officially support `sum(binomial(5, i) for i = 1:5) = 32` different encodings (of combination length 1 to 5) for explicitly discretising a `StateSpaceSet` into a `Vector{Int}`.

The docs are [here](https://juliadynamics.github.io/ComplexityMeasures.jl/previews/PR299/).

## Implications upstream
Since we with the new `CombinationEncoding` have `32` different ways of discretising, in CausalityTools.jl we will have `32` plugin estimators of KL-divergence, `32` plugin estimators of discrete mutual information, `32` plugin estimators for transfer entropy, etc. This doesn't even include more advanced estimation schemes that go beyond plugin estimation.

## New encodings

For the `OrdinalPatterns` outcome space, we explicitly use `encode` to discretise. However, for `AmplitudeAwareOrdinalPatterns`, we don't - because it was intended in the original paper as a correction to the raw counts of outcomes. 

Here, I introduce two new encodings: `AmplitudeEncoding` and `FirstDifferenceEncoding`, which uses average absolute amplitude information, and average first difference magnitude information (relative to some user-defined min/max ranges), to encode state vectors into integer symbols. Since the encodings accept min/max values as mandatory inputs, we can rescale the average amplitude or first difference information to the interval `[0, 1]`, and then discretise this interval. The average amplitude/first difference falls into one of these bins, and the encoded symbol/integer is the number of this bin (enumeration starts from `1`). This is similar to how `GaussianCDF` is implemented, and I use `RectangularBinEncoding` internally for the discretisation onto the binned unit interval.

Roughly speaking, the original amplitude-aware permutation entropy uses a linear combination of these two quantities (although not discretised) as a correction to the counts. Having these encodings as `Encoding` instances allows a user to explicitly discretise data using amplitude/first difference information. This is directly relevant upstream, because I'm using it to discretise data for multi-variable measures such as transfer entropy. 

A similar encoding can be defined based on the `WeightedOrdinalPatterns` too, but isn't included in this PR.

## Combination encodings

These new measures are useful by themselves, but even more powerful when applied together. They may even be combined with the existing encodings. To treat this formally, I have defined a `CombinationEncoding`, which accepts multiple encodings, and constructs a combined encoding by treating the integer symbols (each an integer in the range `[1, 2, ...., n]`, where `n` varies between encodings) as a cartesian index. 

For example, if we use three different encodings, and encode the vector `x = [0.1, 0.5, 0.4, 0.2]`, then we could get the integers `s_a = 2` for `AmplitudeEncoding`, `s_f = 1` for `FirstDifferenceEncoding` and `s_o = 4` for `OrdinalPatternEncoding`. The combined encoding is then `(2, 1, 4)`. Because the total number of outcomes is always known for each encoding, we can map this combined code to a unique integer, and vice versa, by using `LinearIndices` and `CartesianIndices`. 

We can thus apply any combination of encodings to quantify different properties of state vectors. The more encodings that are combined, the larger the number of possible outcomes (integer symbols) are. The total number of possible outcomes is simply the sum of possible outcomes for each input encoding.

## Other changes

- Rearranged the tests for encodings. Tests for each `Encoding` now is in a separate file. Utility tests are moved to a file of its own.
- Implemented `encode(::GaussianCDFEncoding, x::AbstractVector{<:})`, so that `GaussianCDFEncoding` is compatible with `CombinationEncoding`.

This PR is originally part of CausalityTools,  and the encodings here are used in an upcoming paper of mine. But since we now define a public API for encodings and these encodings are completely generic, we should offer everything here.

Datseris commented on src/encoding_implementations/gaussian_cdf.jl:
```diff
@@ -83,3 +83,12 @@ function decode(encoding::GaussianCDFEncoding, i::Int)
     return SVector(lower_interval_bound, prevfloat(lower_interval_bound + 1/c))
 end
+
+# Since the μ and σ are fixed parameters, we can also just encode/decode a state vector
+# element-wise.
+function encode(encoding::GaussianCDFEncoding, x::AbstractVector{<:Real})
+    return encode.(Ref(encoding), x)
+end
+function decode(encoding::GaussianCDFEncoding, x::AbstractVector{<:Real})
+    return decode.(Ref(encoding), x)
```
This to me appears to violate the `encode` interface, which states that the `encode` function returns an integer. this does not return an integer.

kahaaga replied:
> This to me appears to violate the encode interface, which states that the encode function returns an integer. this does not return an integer.

You are right. For an `n` element state vector `x`, there will be `c` different possible integers for each `xi`, so this should also use the cartesian-linear mapping, so that the resulting symbol is an integer

Datseris commented on src/encoding_implementations/amplitude_encoding.jl:
```diff
@@ -0,0 +1,68 @@
+export AmplitudeEncoding
+
+"""
+    AmplitudeEncoding <: Encoding
```
So to get things straight, the `AmplitudeAwareOrdinalPattern` outcome space is the _combination_ encoding of `OrdinalPatternEncoding` and `AmplitudeEncoding`? 

Unfortunately the name `AmplitudeEncoding` is too simple, while this encoding is really complex, and its name doesn't convey what it does. My mind went into value histogram with the name.

More importantly, is this encoding useful in the first place? It hasn't been used in the literature as far as I understand. We should be mindful into not adding too many things in the code base that may have no use on their own.

Datseris replied:
As far as I can tell, this PR does not alter any of the `probabilities` function calls. So, I have to wonder, should we first do some research that shows that all these extra functionality is useful in some way?

Datseris replied:
I just don't see the reason for existence of an encoding without a corresponding outcome space to use it.

kahaaga replied:
> So to get things straight, the AmplitudeAwareOrdinalPattern outcome space is the combination encoding of OrdinalPatternEncoding and AmplitudeEncoding?

When used with `probabilities`, `AmplitudeAwareOrdinalPattern` is (roughly) the combination of `OrdinalPatternEncoding` with a linear combination of `AmplitudeEncoding` and `FirstDifferenceEncoding` dictating the weights. 

> I just don't see the reason for existence of an encoding without a corresponding outcome space to use it.

As stated in the bottom statement in the initial PR comment, the reason is to have complete flexibility in how to discretize data in upstream functions. I am using the new encodings in an upcoming paper to develop some information transfer algorithms that improve on the "naive"/only ordinal pattern based ones. 

They are not yet in any papers, because nobody has used them in this way yet.

Datseris commented on src/encoding_implementations/combination_encoding.jl:
```diff
@@ -0,0 +1,87 @@
+end
+
+function total_outcomes(encoding::CombinationEncoding)
+    return sum(total_outcomes(e) for e in encoding.encodings)
```
Is `sum` correct here? Shouldn't it be `prod`, i.e., the product? From the constructor that's what I gather.

kahaaga replied:
You're right. Well spotted.

Datseris commented on src/encoding_implementations/combination_encoding.jl:
```diff
@@ -0,0 +1,87 @@
+    linear_indices::L
+    cartesian_indices::C
+
+    function CombinationEncoding(encodings::Vararg{<:Encoding, N}) where N
+        ranges = tuple([1:total_outcomes(e) for e in encodings]...)
+        linear_indices = LinearIndices(ranges)
+        cartesian_indices = CartesianIndices(ranges)
+        VE = typeof(encodings)
+        L = typeof(linear_indices)
+        C = typeof(cartesian_indices)
+        return new{VE, L, C}(encodings, linear_indices, cartesian_indices)
+    end
+
+    CombinationEncoding(encodings::Vector{<:Encoding}) = CombinationEncoding(encodings...)
```
I would say these are not inner constructors; they should be defined outside the struct.

Datseris replied:
These constructors should also check that none of the encodings are itself a `CombinationEncoding`.

Datseris commented on src/encoding_implementations/combination_encoding.jl:
```diff
@@ -0,0 +1,87 @@
+    CombinationEncoding(encodings::Vector{<:Encoding}) = CombinationEncoding(encodings...)
+end
+
+function encode(encoding::CombinationEncoding, x::AbstractVector{<:Real})
+    symbols = [encode(e, x) for e in encoding.encodings]
```
I am confused. Why is `x` mandatory to be a vector of reals...?

kahaaga replied:
> I am confused. Why is `x` mandatory to be a vector of reals...?

`CombinationEncoding` is only compatible with state vector inputs, not `Real`s. This should be in the documentation.

Datseris commented on src/encoding_implementations/combination_encoding.jl:
```diff
@@ -0,0 +1,87 @@
+    CombinationEncoding(encodings::Vector{<:Encoding}) = CombinationEncoding(encodings...)
+end
+
+function encode(encoding::CombinationEncoding, x::AbstractVector{<:Real})
+    symbols = [encode(e, x) for e in encoding.encodings]
```
list comprehensions are bad for performance. If we ensure that `encodings` is a tuple, then `map` is better and creates a tuple, and doesn't allocate and is parallelizable as well;

`symbols = map(x -> encode(e, x), encoding.encodings)`

Datseris commented on src/encoding_implementations/combination_encoding.jl:
```diff
@@ -0,0 +1,87 @@
+
+function decode(encoding::CombinationEncoding, ω::Int)
+    cidx = encoding.cartesian_indices[ω]
+    return [decode(e, cidx[i]) for (i, e) in enumerate(encoding.encodings)]
```
similar as above, don't make a vector, use `map` so that a tuple is created.

kahaaga replied:
> similar as above, don't make a vector, use map so that a tuple is created.

I'm not sure I get this comment. `map`ping anything returns a vector:

```julia
julia> map(x -> x, 1:5)
5-element Vector{Int64}:
 1
 2
 3
 4
 5
```

Datseris replied:
Only returns a vector if the collection is a vector. If its a tuple it returns a tuplke and doesnt allocate anything:
```
julia> map(cos, (1,2,3))
(0.5403023058681398, -0.4161468365471424, -0.9899924966004454)

```

so `encodings` must be atuple.

Datseris replied:
also, a tuple is equivalent with the cartesian indices, so we also reduce computations when converting.

Datseris commented on src/encoding_implementations/amplitude_encoding.jl:
```diff
@@ -0,0 +1,68 @@
+    AmplitudeEncoding(minval::Real, maxval::Real; n = 2)
+
+Encoding which discretizes a state vector (an `AbstractVector` of some kind) into its
+"absolute amplitude symbol"  relative to some pre-dedined minimum value `minval` and
```
Absolute amplitude symbol is also weird. This encoding _averages_ the elements of the vector; it does not pick out the one that has the largest amplitude. We should state something like this to guide the reader.

"A vector is encoded based on the relative position the mean of the vector has with respect to a predefined minimum and maximum value".

Datseris commented on src/encoding_implementations/firstdifference_encoding.jl:
```diff
@@ -0,0 +1,82 @@
+    maxval::Real
+    encoder::RectangularBinEncoding
+
+    function FirstDifferenceEncoding(minval::Real, maxval::Real; n = 2)
+        encoder = RectangularBinEncoding(FixedRectangularBinning(0, 1, n + 1))
+        new(n, minval, maxval, encoder)
+    end
```
Convenience constructors should not be inner consttructors. Also, I've noticed that the field `encoder` is not type parameterized.

Datseris commented on src/encoding_implementations/firstdifference_encoding.jl:
```diff
@@ -0,0 +1,82 @@
+export FirstDifferenceEncoding
+
+"""
+    FirstDifferenceEncoding <: Encoding
```
For this encoding I have exactly the same comments as the `AmplitudeEncoding`.

Datseris commented on src/encoding_implementations/utils.jl:
```diff
@@ -0,0 +1,5 @@
+# Normalize `x::Real ∈ [minval, maxval]` to the interval [minval, maxval]. Assumes
+# `minval < maxval`. This needs to be checked at a higher level if this function is called.
+function norm_minmax(x, minval, maxval)
+    return (x - minval) / (maxval - minval)
+end
```
why does this function exist...? Isn't it much simpler to just write `(x - minval) / (maxval - minval)` everywhere in the source code that it uses it? It is definitely much clearer. I'd say to remove this file alltogether.

Datseris commented on test/runtests.jl:
```diff
@@ -27,9 +27,7 @@ testfile(file, testname=defaultname(file)) = @testset "$testname" begin; include
     # include("multiscale/multiscale.jl")
 
     # Various
-    testfile("utils/fasthist.jl")
-    testfile("utils/bin_encoding.jl")
-    testfile("utils/encoding.jl")
```
where did the fasthist tests go to?

Datseris reviewed (commented):
We need to discuss a bit the usefulness of adding things that are not used in the "main" API of outcome spaces, and have not been used in the literature explicitly either.

kahaaga commented:
> I just don't see the reason for existence of an encoding without a corresponding outcome space to use it.

We used to have a `symbolize` function here before,  which had the role of *explicitly* symbolizing some `StateSpaceSet`. 
We decided to remove it when formalizing the encoding API. This `symbolize` function is not needed for the purposes of estimating single-variable information measures, as we do in this package. However, it is needed in upstream discrete multi-variable estimators. 

With the proposal here, I can implement a completely generic `symbolize` upstream. Currently, I have to write special implementations upstream for every `OutcomeSpace` (which now is basically copy-pasting a lot of code from ComplexityMeasures).  

Just as the various `ProbabilitiesEstimator`s and `DiscreteInfoEstimator`s give us `X` number of ways to compute e.g. `Shannon` entropy, this generic encoding API will allow `Y` number of ways to estimate multi-variable information measures upstream, many of which have not been explored (in my case, simply due to lack of dedicated software for doing so).

A possible resolution to your comment is to define a generic `CombinationOutcomeSpace` that uses `CombinationEncoding` under the hood. That links all the underlying `Encodings` to the estimation of probabilities and information measures. That would actually be nice to have, I think.

kahaaga commented:
> this pr makes me think of two things:
> 
> 
> 
> 1) `WeightedOrdinalPatterns` (previously `WeightedSymbolicPermutation`) isn't an outcome space, right?
> 
> 2) Is there really a reason to have **both** `OutcomeSpace` and `Encoding`? shouldn't we just have outcome spaces that **support** the `encode` function, and some outcome spaces that simply do not support `encode`, such as `PowerSpectrum`? All count-based outocme spaces support `encode`. This would simplify things a lot: we would drop a large amount of the type hierarchy, which is everything related to encodings.
> 
> 3) Yes, the existence of `CombinationOutcomeSpace` makes perfect sense to me.

I think your suggestion in 2) is an excellent idea. I'll sketch a variant of this PR that does so. 

And yes, you are correct about 1).

Datseris commented:
> I'll sketch a variant of this PR that does so.

perhaps we should keep things clean. Let's leave this PR as is now: contributing new encodings.

 A different PR should merge the two concepts.

Datseris commented:
The new pr shold also make `CombinationOutcomeSpace `. So for now just focus on addressing the standing comments.

kahaaga commented:
> `WeightedOrdinalPatterns` (previously `WeightedSymbolicPermutation`) isn't an outcome space, right?

It isn't an outcome space now, but it is trivial to define one that uses the same information (variance of the state vectors) when discretizing. This is completely analogous to the two new proposed encodings (which will become outcome spaces).

kahaaga commented:
"Weighted relative frequencies" is the term that Fadlallah et al. use when they describe the weighting for the weighted permutation entropy, so strictly speaking this is  justa modified/weighted `RelativeAmount` probabilities estimator.
With this way of thinking, for the new PR, we should probably define a `WeightedRelativeAmount` probabilities estimator. 

But we need to think a bit about how to incorporate the weights with the `WeightedRelativeAmount` estimator. Do the weights go as a field to the estimator, for example like the following?

I can't remember the exact details of the implementation at the moment, but something like this:

```julia
"""
    WeightedRelativeAmount <: ProbabilitiesEstimator
    WeightedRelativeAmount(outcomemodel::OutcomeSpace, wts::AbstractVector{<:Real})
    
Like [`RelativeAmount`](@ref), but *some text explaining how weights come into play*.
"""
struct WeightedRelativeAmount{O <: OutcomeSpace, T} <: ProbabilitiesEstimator
    outcomemodel::O
    weights::AbstractVector{<:T}
end

function probabilities(est::WeightedRelativeAmount, x)
     symbols = encode.(Ref(est.outcomemodel), x)
     return Probabilities(fasthist(symbols, weights)) #weighted, normalized histogram 
end

# Can't remember the signature right now, but something like
function weighted_permutation_entropy(x::StateSpaceSet{m}; kwargs....) where m
    emb = embed(x, tau, m)
    est = WeightedRelativeAmount(OrdinalPatterns(m), weights_from_variance.(emb))
    return entropy(PlugIn(Shannon(), probabilities(est, x)))
end
```

The same goes for `AmplitudeAwareOrdinalPatterns` - the weights are just calculated differently.

This feels a bit cumbersome, but could work.

Datseris commented:
id like to keep new features and new things to a minimum so i'd say "no" to a new probabilities estimator. let's try to add less things rather than more. Just keep it as is: an encoding / outcome space.

Datseris commented:
and, in any case, the mathematics of probability spaces do not have "weights". This is a thing of statistics, which we do not do here.

kahaaga commented:
I think I addressed most of the main comments now.

TODO:
- [ ] come up with a more descriptive name for `AmplitudeEncoding`
- [ ] come up with a more descriptive name for `FirstDifferenceEncoding`

kahaaga commented:
> Just keep it as is: an encoding / outcome space.

Do you mean that `WeightedOrdinalPatterns` and `AmplitudeAwareOrdinalPatterns` stay like they are as `OutcomeSpace`s, but do not have implementations for `encode`? 

> This is a thing of statistics, which we do not do here.

This is just semantics, but I'd argue we almost exclusively do statistics in this package, since all we deal with is [estimation](https://en.wikipedia.org/wiki/Estimator) and the computation of various statistics (functions of data). It might not be though of as such in the NLD community, but most of our codebase falls well within the realm of statistics. 

> and, in any case, the mathematics of probability spaces do not have "weights". 

Again just semantics, but the probability function ([probability measure](https://en.wikipedia.org/wiki/Probability_measure)) on a probability space is extremely generic, and I don't immediate see any reason why weights couldn't be incorporated in such a function (unless it can be proven otherwise). 

In any case, when *estimating* probabilities, weights are definitely a thing, and all the current `ProbabilitiesEstimator`s use some form of weighting: `AddConstant` uses additive weighting, `BayesianRegularization` uses a combination of additive and multiplicative weighting, etc. 

However, I'll think a bit more about this `WeightedRelativeAmount` and open a PR later if relevant. If designed cleverly, it could be used for incorporating prior knowledge from the user into the probabilities estimates.

Datseris commented:
so this is good to go?

Datseris commented:
or, good for a review?

Datseris commented:
yes, I'll try to do it asap.

Datseris commented on src/encoding_implementations/combination_encoding.jl:
```diff
@@ -0,0 +1,97 @@
+# To do so, we will use a combination of first-difference encoding, amplitude encoding,
+# and ordinal pattern encoding.
+
+encodings = [
+    RelativeFirstDifferenceEncoding(0, 1; n = 2),
+    RelativeMeanEncoding(0, 1; n = 5),
+    OrdinalPatternEncoding(3) # x is a three-element vector
+    ]
```
```suggestion
encodings = (
    RelativeFirstDifferenceEncoding(0, 1; n = 2),
    RelativeMeanEncoding(0, 1; n = 5),
    OrdinalPatternEncoding(3) # x is a three-element vector
    )
```
using a tuple is more efficient.

Datseris commented on src/encoding_implementations/combination_encoding.jl:
```diff
@@ -0,0 +1,97 @@
+end
+CombinationEncoding(encodings::Vector{<:Encoding}) = CombinationEncoding(encodings...)
+
+# We could in principle allow any `x` here, but not all encodings support encoding
+# single numbers. In particular, the `RelativeFirstDifferenceEncoding` isn't even defined
+# for single numbers, and `OrdinalPatternEncoding` also isn't defined for single numbers.
+# Therefore, we enforce vector-valued input with encoding.
+function encode(encoding::CombinationEncoding, x::AbstractVector{<:Real})
```
I don't get this argument. If anything, you shouldn't parameterize `x` at all. The encodings themselves would throw a method error when you call `encode` below.

Datseris replied:
Also, rename `x` to `χ`.

Datseris commented on src/encoding_implementations/combination_encoding.jl:
```diff
@@ -0,0 +1,97 @@
+end
+
+function total_outcomes(encoding::CombinationEncoding)
+    return prod(total_outcomes(e) for e in encoding.encodings)
```
```suggestion
    return prod(total_outcomes.(encoding.encodings))
```

Datseris commented on src/encoding_implementations/gaussian_cdf.jl:
````diff
@@ -54,32 +75,91 @@ julia> decode(encoding, 3)
  0.4
  0.6
 ```
+
+One can also encode the entire vector as an integer.
+
+```jldoctest
+julia> using ComplexityMeasures, Statistics
+
+julia> x = [0.1, 0.4, 0.7, -2.1, 8.0];
+
+julia> μ, σ = mean(x), std(x); encoding = GaussianCDFEncoding(x; μ, σ, c = 2)
+GaussianCDFEncoding(m=5; c=2, μ=1.42, σ=3.840182287340016)
+
+julia> symbol = encode(encoding, x)
+17
+
+julia> decode(encoding, symbol)
+5-element Vector{SVector{1, Float64}}:
+ [0.0]
+ [0.0]
+ [0.0]
+ [0.0]
+ [0.5000000000000001]
+```
````
I would remove this example for brevity. If it is useful somewhere else, like an actual example, we can put it in the examples in the docs.

Datseris commented on src/encoding_implementations/gaussian_cdf.jl:
```diff
@@ -54,32 +75,91 @@ julia> decode(encoding, 3)
+    binencoder::R # RectangularBinEncoding
+
+    # The input `m` restricts what length the input scalar/vector can be.
+    function GaussianCDFEncoding(m::Int = 1; μ::T, σ::T, c::Int = 3) where T
```
Just FYI, `GaussianCDFEncoding` is created now in a type unstable manner, as the input value `m` decides the type `m` later. This means that if code constructs and calls it in a single function, that is type unstable code.

There is no way to fix this while still allowing `m` as an input.

I think this performance defecit doesn't matter much unless you are looping over calls to `encode(GaussianCDF, χ)`.

kahaaga replied:
> I think this performance defecit doesn't matter much unless you are looping over calls to encode(GaussianCDF, χ).

Then we probably should resolve this here, because that is precisely the intention of allowing vector-valued inputs ( upstream).

Datseris replied:
You can resolve this by making the constructor `GaussianCDFEncoding{m}(; μ...)` So that `m` is given directly as a type parameter by the user.

Datseris replied:
@kahaaga you haven't addressed this comment. You should make this change and then merge. You should also create the default constructor `GaussianCDFEncoding(; μ ...)` that assumes `m=1`.

Datseris commented on src/encoding_implementations/gaussian_cdf.jl:
```diff
@@ -54,32 +75,91 @@ julia> decode(encoding, 3)
 end
+GaussianCDFEncoding(x::AbstractVector; kwargs...) = GaussianCDFEncoding(length(x); kwargs...)
+
+function Base.show(io::IO, e::GaussianCDFEncoding{m, T, L, C}) where {m, T, L, C}
+    c, μ, σ = e.c, e.μ, e.σ
+    print(io, "GaussianCDFEncoding(m=$m; c=$c, μ=$μ, σ=$σ)")
+end
```
We should have a generic `show` method for everything instead of adding `show` here. This `show` method is caught by the test coverage. I would argue, let's remove adding any `show` here and add them all in a generic `show` PR. CF #286

Datseris commented on src/encoding_implementations/gaussian_cdf.jl:
```diff
@@ -54,32 +75,91 @@ julia> decode(encoding, 3)
-total_outcomes(encoding::GaussianCDFEncoding) = encoding.c
+function total_outcomes(encoding::GaussianCDFEncoding{m}) where m
+    c = encoding.c
+    return prod(c for i = 1:m)
```
```suggestion
    return c^m
```

Datseris commented on test/encodings/encodings/combination_encoding.jl:
```diff
@@ -0,0 +1,17 @@
+
+e_combo = CombinationEncoding(es)
+
+symbol = encode(e_combo, x)
+@test symbol isa Int
+d = decode(e_combo, symbol)
+@test d isa AbstractVector
+@test length(d) == length(es)
```
These aren't really correctness tests, they are only API test. Why don't we have a correctness test here? Since there are analytic t ests for OrdinalPatternEncoding, and hte behavior of `RelativeMeanEncoding` is straightforward to interpret, please add an analytic test that tests exactly that an encoder integer is exactly 15 or whatever number it should be.

kahaaga commented:
> @kahaaga you haven't addressed this comment. You should make this change and then merge. You should also create the default constructor GaussianCDFEncoding(; μ ...) that assumes m=1.

I'm not finished with addressing the reviews yet. Working on it as we speak 👍 

(If you got a notification for a PR review, it should have been for #302).
